compiler/rustc_typeck/src/check/writeback.rs

```diff
@@ -192,6 +192,27 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         }
     }
 
+    // (ouz-a 1005988): Normally `[T] : std::ops::Index<usize>` should be normalized
+    // into [T] but currently `Where` clause stops the normalization process for it,
+    // here we compare types of expr and base in a code without `Where` clause they would be equal
+    // if they are not we don't modify the expr, hence we bypass the ICE
+    fn is_builtin_index(
+        &mut self,
+        typeck_results: &TypeckResults<'tcx>,
+        e: &hir::Expr<'_>,
+        base_ty: Ty<'tcx>,
+        index_ty: Ty<'tcx>,
+    ) -> bool {
+        if let Some(elem_ty) = base_ty.builtin_index() {
+            let exp_ty = typeck_results.expr_ty_adjusted(e);
+            let resolved_exp_ty = self.resolve(exp_ty, &e.span);
+
+            elem_ty == resolved_exp_ty && index_ty == self.fcx.tcx.types.usize
+        } else {
+            false
+        }
```

Comment:
Actually... it might be wrong to use `expr_ty_adjusted` here. The un-adjusted expr should be `&T`, so maybe try:

```suggestion
        let Some(expected_elem_ty) = base_ty.builtin_index() else { return false; };
        let Some(exp_ty) = typeck_results.expr_ty_opt(e) else { return false; };
        let ty::Ref(_, found_elem_ty, _) = self.resolve(exp_ty, &e.span).kind();
        expected_elem_ty == found_elem_ty && index_ty == self.fcx.tcx.types.usize
```